model/context.go

```diff
@@ -55,10 +58,67 @@ type Url struct {
 	Fragment *string
 }
 
-// Page consists of Url string and referer
+func ParseURL(original, hostname string) *URL {
+	original = truncate(original)
+	url, err := url.Parse(original)
+	if err != nil {
+		return &URL{Original: &original}
+	}
+	if url.Scheme == "" {
+		url.Scheme = "http"
+	}
+	if url.Host == "" {
+		url.Host = hostname
+	}
+	full := truncate(url.String())
+	out := &URL{
+		Original: &original,
+		Scheme:   &url.Scheme,
+		Full:     &full,
+	}
+	if path := truncate(url.Path); path != "" {
+		out.Path = &path
+	}
+	if query := truncate(url.RawQuery); query != "" {
+		out.Query = &query
+	}
+	if fragment := url.Fragment; fragment != "" {
+		out.Fragment = &fragment
+	}
+	host, port, err := net.SplitHostPort(url.Host)
+	if err != nil {
+		host = truncate(url.Host)
+		port = ""
+	}
```

Comment:
I think you probably want to use url.Hostname() and url.Port(), rather than SplitHostPort.

The rules are a little different for splitting url.Host, e.g. in url.Host it's required to have an IPv6 address enclosed in square brackets even if there's no port specified, whereas SplitHostPort requires there to be a port if square brackets are found.